daemon/containerd/service.go

```diff
@@ -0,0 +1,369 @@
+package containerd
+
+import (
+	"context"
+	"io"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/platforms"
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/backend"
+	"github.com/docker/docker/api/types/filters"
+	imagetype "github.com/docker/docker/api/types/image"
+	registrytypes "github.com/docker/docker/api/types/registry"
+	"github.com/docker/docker/builder"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/daemon/images"
+	"github.com/docker/docker/errdefs"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/layer"
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+// ImageService implements daemon.ImageService
+type ImageService struct {
+	client *containerd.Client
+}
+
+// NewService creates a new ImageService.
+func NewService(c *containerd.Client) *ImageService {
+	return &ImageService{
+		client: c,
+	}
+}
+
+// PullImage initiates a pull operation. image is the repository name to pull, and
+// tag may be either empty, or indicate a specific tag to pull.
+func (cs *ImageService) PullImage(ctx context.Context, image, tag string, platform *ocispec.Platform, metaHeaders map[string][]string, authConfig *types.AuthConfig, outStream io.Writer) error {
+	var opts []containerd.RemoteOpt
+	if platform != nil {
+		opts = append(opts, containerd.WithPlatform(platforms.Format(*platform)))
+	}
+	ref, err := reference.ParseNormalizedNamed(image)
+	if err != nil {
+		return errdefs.InvalidParameter(err)
+	}
+
+	if tag != "" {
+		// The "tag" could actually be a digest.
+		var dgst digest.Digest
+		dgst, err = digest.Parse(tag)
+		if err == nil {
+			ref, err = reference.WithDigest(reference.TrimNamed(ref), dgst)
+		} else {
+			ref, err = reference.WithTag(ref, tag)
+		}
+		if err != nil {
+			return errdefs.InvalidParameter(err)
+		}
+	}
+
+	_, err = cs.client.Pull(ctx, ref.String(), opts...)
+	return err
+}
+
+// Images returns a filtered list of images.
+func (cs *ImageService) Images(ctx context.Context, opts types.ImageListOptions) ([]*types.ImageSummary, error) {
```

Comment:
Oh man, reimplementing some of these image list filter options is going to be interesting. :sweat_smile:

(Might make more sense to redesign the interface?  No idea how similar the implementations will be or what filter options containerd's API supports directly.)

Maybe another place a `TODO` comment would make sense if it's not already tracked elsewhere. :eyes:

Reply:
I [already started extracting things from the image service that have no reason to be there](https://github.com/moby/moby/pull/43680), we will definitely have to clean this up, I don't want to implement these things twice :)